instrumentation/aws-java-sdk-sqs-2.1.0/src/main/java/software/amazon/awssdk/services/sqs/DefaultSqsClient_Instrumentation.java

```diff
@@ -7,43 +7,259 @@
+import java.util.Map;
+
 @Weave(type = MatchType.ExactClass, originalName = "software.amazon.awssdk.services.sqs.DefaultSqsClient")
 class DefaultSqsClient_Instrumentation {
 
     @Trace
     public SendMessageBatchResponse sendMessageBatch(SendMessageBatchRequest sendMessageBatchRequest) {
-        MessageProduceParameters messageProduceParameters = MetricUtil.generateExternalProduceMetrics(sendMessageBatchRequest.queueUrl());
+        List<SendMessageBatchRequestEntry> updatedEntries = new ArrayList<>();
+        for (SendMessageBatchRequestEntry entry : sendMessageBatchRequest.entries()) {
+            SQSBatchRequestHeaders headers = new SQSBatchRequestHeaders(entry);
+            NewRelic.getAgent().getTransaction().insertDistributedTraceHeaders(headers);
+            updatedEntries.add(headers.updatedEntry());
+        }
+        sendMessageBatchRequest = sendMessageBatchRequest.toBuilder().entries(updatedEntries).build();
+
+        MessageProduceParameters messageProduceParameters = SqsV2Util.generateExternalProduceMetrics(sendMessageBatchRequest.queueUrl());
         NewRelic.getAgent().getTracedMethod().reportAsExternal(messageProduceParameters);
+
         return Weaver.callOriginal();
     }
 
     @Trace
     public SendMessageResponse sendMessage(SendMessageRequest sendMessageRequest) {
-        MessageProduceParameters messageProduceParameters = MetricUtil.generateExternalProduceMetrics(sendMessageRequest.queueUrl());
+        SQSRequestHeaders headers = new SQSRequestHeaders(sendMessageRequest);
+        NewRelic.getAgent().getTransaction().insertDistributedTraceHeaders(headers);
+        sendMessageRequest = headers.getUpdatedRequest();
+
+        MessageProduceParameters messageProduceParameters = SqsV2Util.generateExternalProduceMetrics(sendMessageRequest.queueUrl());
         NewRelic.getAgent().getTracedMethod().reportAsExternal(messageProduceParameters);
+
         return Weaver.callOriginal();
     }
 
     @Trace
     public ReceiveMessageResponse receiveMessage(ReceiveMessageRequest receiveMessageRequest) {
-        MessageConsumeParameters messageConsumeParameters = MetricUtil.generateExternalConsumeMetrics(receiveMessageRequest.queueUrl());
+        ArrayList<String> updatedMessageAttrNames = new ArrayList<>(receiveMessageRequest.messageAttributeNames());
+        Collections.addAll(updatedMessageAttrNames, SqsV2Util.DT_HEADERS);
+        receiveMessageRequest = receiveMessageRequest.toBuilder().messageAttributeNames(updatedMessageAttrNames).build();
+
+        MessageConsumeParameters messageConsumeParameters = SqsV2Util.generateExternalConsumeMetrics(receiveMessageRequest.queueUrl());
         NewRelic.getAgent().getTracedMethod().reportAsExternal(messageConsumeParameters);
+
         return Weaver.callOriginal();
     }
+
+    /*
+     * Headers are used as inner classes to avoid class loading errors when running instrumentation tests
+     * */
+
+    public class SQSRequestHeaders implements Headers {
```

Comment:
Move this inner class to its own file and use the same on the async client.

Reply:
Problem here is when I did that, I got class-loading errors when running instrumentation tests. I suspect its some magic the SDK is using with classloaders that did not mesh well with the agent. This was the one way I got it to work.

Reply:
You need to add the package that has the header classes to the @InstrumentationTestConfig
In this case it would look like:
```
@InstrumentationTestConfig(includePrefixes = { "software.amazon.awssdk.services.sqs", "com.newrelic.utils" }, configName = "dt_enabled.yml")
```

Reply:
Moved the header classes

Reply:
Thanks for the suggestion. I moved the header classes and the tests pass now.